### ISSUE TITLE
Type stubs for rosbag2_py

### ISSUE THREAD

**Issue by MichaelFishmanBDAII** [ros2/rosbag2#1459]
## Description

I'd like type stubs for `rosbag2_py`, so that I can get autocomplete/static analysis.

## Completion Criteria

** What needs to be true before we can call this "Done"? Bullet lists are appropriate. **

- The maximal version: All classes/methods/functions exposed by `rosbag2_py` have full type annotations. A static analysis tool like mypy/pylance can tell parse these type annotations.

## Implementation Notes / Suggestions
Stubs can be generated with `pybind11-stubgen`:

```sh
pip install pip install pybind11-stubgen

pybind11-stubgen rosbag2_py
```
After doing this, making a `setup.py`, and installing the newly created `rosbag2_py-stubs` library, we get some type information, though the function argument and return types are incomplete:

![image](https://github.com/ros2/rosbag2/assets/142259135/fe0e943f-bb25-41e9-9f22-8b1fd13493cd)

The output from the `pybind11-stubgen` command is

```
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialCompressionReader : Invalid expression 'rosbag2_storage::TopicMetadata'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialCompressionReader : Invalid expression 'rosbag2_storage::StorageOptions'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialCompressionReader : Invalid expression 'rosbag2_cpp::ConverterOptions'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialCompressionReader : Invalid expression 'rosbag2_storage::StorageFilter'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialReader : Invalid expression 'rosbag2_storage::TopicMetadata'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialReader : Invalid expression 'rosbag2_storage::StorageOptions'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialReader : Invalid expression 'rosbag2_cpp::ConverterOptions'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialReader : Invalid expression 'rosbag2_storage::StorageFilter'
pybind11_stubgen - [WARNING] Raw C++ types/values were found in signatures extracted from docstrings.
Please check the corresponding sections of pybind11 documentation to avoid common mistakes in binding code:
 - https://pybind11.readthedocs.io/en/latest/advanced/misc.html#avoiding-cpp-types-in-docstrings
 - https://pybind11.readthedocs.io/en/latest/advanced/functions.html#default-arguments-revisited
 ```

I expect that addressing the issues in this output would get us the missing type information.

## Testing Notes / Suggestions

** All features in this project need tests. Please give some input on cases that will need to be tested - and how the testing might be implemented. **

A test could run `mypy` on an example python script using `rosbag2_py`. I'm not sure what `mypy` options would need to be set to verify that all the `rosbag2_py` objects are properly typed.

**Comment by MichaelOrlov** [comment]
@emersonknapp Do you have thoughts about this proposal?

**Comment by MichaelFishmanBDAII** [comment]
Currently, the type annotations for methods are generic. Eg. `SequentialReader.read_next` has the return type of `tuple`:

```c++
pybind11::tuple read_next()
```

Ideally, it would have the return type `tuple[str, bytes, int]`. 

pybind's [test_python_types.cpp](https://github.com/virtuald/pybind11/blob/7830e8509f2adc97ce9ee32bf99cd4b82089cc4c/tests/test_python_types.cpp#L129) has examples showing how to specify these more specific types.


```c++
    /* pybind automatically translates between C++11 and Python tuples */
    std::tuple<int, std::string, bool> tuple_passthrough(std::tuple<bool, std::string, int> input) {
        return std::make_tuple(std::get<2>(input), std::get<1>(input), std::get<0>(input));
    }
```

**Comment by r7vme** [comment]
@MichaelOrlov should we reopen this issue or create a new one for addressing issues that pybind11-stubgen found?

```
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialCompressionReader : Invalid expression 'rosbag2_storage::TopicMetadata'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialCompressionReader : Invalid expression 'rosbag2_storage::StorageOptions'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialCompressionReader : Invalid expression 'rosbag2_cpp::ConverterOptions'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialCompressionReader : Invalid expression 'rosbag2_storage::StorageFilter'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialReader : Invalid expression 'rosbag2_storage::TopicMetadata'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialReader : Invalid expression 'rosbag2_storage::StorageOptions'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialReader : Invalid expression 'rosbag2_cpp::ConverterOptions'
pybind11_stubgen - [  ERROR] In rosbag2_py._reader.SequentialReader : Invalid expression 'rosbag2_storage::StorageFilter'
pybind11_stubgen - [WARNING] Raw C++ types/values were found in signatures extracted from docstrings.
Please check the corresponding sections of pybind11 documentation to avoid common mistakes in binding code:
 - https://pybind11.readthedocs.io/en/latest/advanced/misc.html#avoiding-cpp-types-in-docstrings
 - https://pybind11.readthedocs.io/en/latest/advanced/functions.html#default-arguments-revisited

```

**Comment by MichaelOrlov** [comment]
@r7vme Hold on. But how the pybind11-stubgen found these issues if we didn't change those files from the last run?

Certainly need to open a new issue or PR.

**Comment by r7vme** [comment]
@MichaelOrlov  I just copy pasted output from the above. We did not break anyting yet :) In any case, I think we can try to address those issues.

**Comment by r7vme** [comment]
Can we reopen this issue?

**Comment by MichaelOrlov** [comment]
Ok. I reopen this issue.

**Comment by MichaelOrlov** [comment]
@r7vme Am I understand correctly that we already added py stubs in the scope of the #1569 and can close this issue now?

**Comment by r7vme** [comment]
Yeah, let's close this issue, major issue is fixed.
